Add docs for 💯 TypeScript support (#723)
redwoodjs/redwoodjs.com: docs/typescript.md

# Typescript

> Redwood comes with full TypeScript support out of the box, and you don't have to give up any of the convenience in Cells, Router or generators, while still enjoying all the benefits of a type-safe codebase.

Reviewer: - figured we could do this instead of enumerating the conveniences
```suggestion
Redwood comes with full TypeScript support out of the box, and you don't have to give up any of the conveniences that Redwood offers to enjoy all the benefits of a type-safe codebase. You can use TypeScript and have DX too. 
```

## Starting a TypeScript Redwood project
You can use the `--typescript` flag on create-redwood-app to generate a project with TypeScript configured
```shell
yarn create redwood-app --typescript /PATH/TO/FOLDER
```

## Converting an existing JS project to Typescript
If you already have a Redwood app, but want to configure it for TypeScript, you can use our setup command

```
yarn rw setup tsconfig
```
Remember you don't _need_ to convert all your files to TypeScript, you can always do it incrementally. Start by renaming your files from `.js` to `.ts` or `.tsx`

Having issues with automatic setup? See instructions for manual setup

<details>
<summary>Manually setup TypeScript</summary>
This is what the setup command does for you

**API**

1. Create a `./api/tsconfig.json` file:

```shell
touch api/tsconfig.json

> ```

Reviewer: It's hard to use github's suggestion feature here because of the backticks, but can you add a `<br />` after this? The spacing is just a little too tight (it's more obvious in the screenshot taken in another comment)

> 

Reviewer 2: ```suggestion

<br />

```

> 2. Now copy and paste the latest config from the Redwood template: [`api/tsconfig.json`](https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/tsconfig.json) into the empty file

Reviewer: - kind of hard to see that this is a link
```suggestion
2. Now copy and paste the latest config from the Redwood template [api/tsconfig.json](https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/tsconfig.json) file here
```

**WEB**

1. Create a `./api/tsconfig.json` file:

```shell
touch web/tsconfig.json
```

> 2. Now copy and paste the latest config from the Redwood template: [`web/tsconfig.json`](https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/web/tsconfig.json) into the empty file

Reviewer: - kind of hard to see that this is a link
```suggestion
2. Now copy and paste the latest config from the Redwood template [web/tsconfig.json](https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/web/tsconfig.json) file here
```

You should now have type definitions&mdash;you can rename your files from `.js` to `.ts`, and the files that contain JSX to `.tsx`.
</details>

## Sharing Types between sides

> For your shared types we need to do a few things:

Reviewer: - add comma
```suggestion
For your shared types, we need to do a few things:
```

```

## Running type checks

Redwood uses Babel to transpile your TypeScript - which is why you are able to incrementally convert your project from JS to TS. However, it also means that just doing a build won't show you errors that the TypeScript compiler finds in your project <br/> <br/> That's why we have the handy `redwood typecheck` command!

To check your TypeScript project for errors, run
```
yarn rw typecheck
```
This will run `tsc` on all the sides in your project, and make sure all the generated types are generated first, including Prisma.

### Check then build, on CI
> **Tip!**<br/>
> You don't need to build your project to run `rw typecheck`

If your project is fully TypeScript, it might be useful to add typechecks before you run the deploy command in your CI.

For example, if you're deploying to Vercel, you could add a `build:ci` script to your `package.json`
```diff
"scripts": {
  .
  .
+  "build:ci": "yarn rw typecheck && yarn rw test && yarn rw deploy vercel",
}
```
Configure your project's build command to be `yarn build:ci` - and you even have your type checks and tests run whenever a pull request is opened!

## Auto generated types

> Redwood's CLI will automatically generate types for you, including types for your graphql queries, but also so that you get types for your named routes, Cells, scenarios and tests.

Reviewer: - case GraphQL like this
- present tense
- including / but also read a little awkward—@dac09 did I keep the same spirit?
```suggestion
Redwood's CLI automatically generates types for you, which not only includes types for your GraphQL queries, but also for your named routes, Cells, scenarios and tests.
```

> When you run `yarn rw dev`, the CLI watches for file changes and will automatically trigger the type generator.

Reviewer: - present tense
```suggestion
When you run `yarn rw dev`, the CLI watches for file changes and automatically triggers the type generator.
```

> To trigger type generation, you can run

Reviewer: - add colon
```suggestion
To trigger type generation, you can run:
```

If you have an outdated tsconfig, and would like to replace it (i.e. overwrite it), use the setup command with the force flag

```
yarn rw setup tsconfig --force
```

## Running type checks

Redwood uses Babel to transpile your TypeScript - which is why you are able to incrementally convert your project from JS to TS. However, it also means that just doing a build won't show you errors that the TypeScript compiler finds in your project <br/> <br/> That's why we have the handy `redwood typecheck` command!

To check your TypeScript project for errors, run
```
yarn rw typecheck
```
This will run `tsc` on all the sides in your project, and make sure all the generated types are generated first, including Prisma.

### Check then build, on CI
> **Tip!**<br/>
> You don't need to build your project to run `rw typecheck`

If your project is fully TypeScript, it might be useful to add typechecks before you run the deploy command in your CI.

For example, if you're deploying to Vercel, you could add a `build:ci` script to your `package.json`
```diff
"scripts": {
  .
  .

> +  "build:ci": "yarn rw typecheck && yarn rw test && yarn rw deploy vercel",

Author: ```suggestion
+  "build:ci": "yarn rw typecheck && yarn rw test --no-watch && yarn rw deploy vercel",
```

## Starting a TypeScript Redwood project
You can use the `--typescript` flag on create-redwood-app to generate a project with TypeScript configured:
```shell
yarn create redwood-app --typescript /PATH/TO/FOLDER
```

## Converting an existing JS project to Typescript
If you already have a Redwood app, but want to configure it for TypeScript, you can use our setup command:

```
yarn rw setup tsconfig
```
Remember you don't _need_ to convert all your files to TypeScript, you can always do it incrementally. Start by renaming your files from `.js` to `.ts` or `.tsx`

Having issues with automatic setup? See instructions for manual setup below:

<details>
<summary>Manually setup TypeScript</summary>

This is what the setup command does for you step by step:

**API**

1. Create a `./api/tsconfig.json` file:

```shell
touch api/tsconfig.json
```

<br />

2. Now copy and paste the latest config from the Redwood template [api/tsconfig.json](https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/tsconfig.json) file here

**WEB**

1. Create a `./api/tsconfig.json` file:

```shell
touch web/tsconfig.json
```

> 

Reviewer 2: ```suggestion

<br />

```
